src/coding-guidelines/types-and-traits.rst

```diff
@@ -63,3 +64,225 @@ Types and Traits
              current.checked_add(velocity).expect("Position calculation overflowed")
          }
 
+.. guideline:: Do Not Depend on Function Pointer Identity Across Crates
+    :id: gui_QbvIknd9qNF6 
+    :category: required
+    :status: draft
+    :release: unclear-latest
+    :fls: fls_1kg1mknf4yx7
+    :decidability: decidable
+    :scope: system
+    :tags: surprising-behavior
+
+    Do not rely on the equality or stable identity of function pointers originating from different crates or that may be inlined,
+    duplicated, or instantiated differently across compilation units, codegen units, or optimization profiles.
+
+    Avoid assumptions about low-level metadata (such as symbol addresses) unless explicitly guaranteed by the Ferrocene Language Specification (FLS).
+    Function address identity is not guaranteed by Rust and must not be treated as stable.
+    Rust’s ``fn`` type is a zero-sized function item promoted to a function pointer, whose address is determined by the compiler backend.
+    When a function resides in a different crate, or when optimizations such as inlining,
+    link-time optimization, or codegen-unit partitioning are enabled,
+    the compiler may generate multiple distinct code instances for the same function or alter the address at which it is emitted.
+
+    Consequently, the following operations are not reliable:
+
+    - Comparing function pointers for equality (``fn1 == fn2``)
```

Comment:
This is reliable for functions which are `#[no_mangle]`

Reply:
@programmerjake i think this has been refuted over in [#general > UB in Safe Rust @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/122651-general/topic/UB.20in.20Safe.20Rust/near/562267154)